pkg/api/v1/types.go

```diff
@@ -710,6 +725,22 @@ type Container struct {
 	TTY   bool `json:"tty,omitempty" description:"Whether this container should allocate a TTY for itself, also requires 'stdin' to be true; default is false"`
 }
 
+func (ctr *Container) ApplyDefaults() {
```

Comment:
Declarative is valid for some fraction of defaults/validation but not all of it.  The problems:

* the validation/defaulting logic gets spread across the tags and the custom methods (though there would be less methods)

* not easy to define new validation functions and use them (e.g. "0 < x < 65536" is less descriptive than "isValidPort()".  

* some of the defaulting/validation is visible to clients of the structs and some is not (maybe?)

* even more magical than before


I think the value that methods bring is the obviousness.  It implements an interface - just like every other Go code.

I'll take a look at declarative approaches - I saw a lib to do some of this, will investigate.